src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs

```diff
@@ -17,17 +18,42 @@ public class DefaultModelMetadataProvider : IModelMetadataProvider
         private readonly TypeCache _typeCache = new TypeCache();
         private readonly Func<ModelMetadataIdentity, ModelMetadataCacheEntry> _cacheEntryFactory;
         private readonly ModelMetadataCacheEntry _metadataCacheEntryForObjectType;
+        private readonly ModelBindingMessageProvider _messageProvider;
 
         /// <summary>
         /// Creates a new <see cref="DefaultModelMetadataProvider"/>.
         /// </summary>
         /// <param name="detailsProvider">The <see cref="ICompositeMetadataDetailsProvider"/>.</param>
+        /// <remarks>Use other constructor overload. This is provided for back compatibility and a few tests.</remarks>
         public DefaultModelMetadataProvider(ICompositeMetadataDetailsProvider detailsProvider)
+            : this(detailsProvider, new ModelBindingMessageProvider())
         {
-            DetailsProvider = detailsProvider;
+        }
 
-            _cacheEntryFactory = CreateCacheEntry;
+        /// <summary>
+        /// Creates a new <see cref="DefaultModelMetadataProvider"/>.
+        /// </summary>
+        /// <param name="detailsProvider">The <see cref="ICompositeMetadataDetailsProvider"/>.</param>
+        /// <param name="optionsAccessor">The accessor for <see cref="MvcOptions"/>.</param>
+        public DefaultModelMetadataProvider(
+            ICompositeMetadataDetailsProvider detailsProvider,
+            IOptions<MvcOptions> optionsAccessor)
+            : this(detailsProvider, GetMessageProvider(optionsAccessor))
+        {
+        }
 
+        private DefaultModelMetadataProvider(
+            ICompositeMetadataDetailsProvider detailsProvider,
+            ModelBindingMessageProvider messageProvider)
+        {
+            if (detailsProvider == null)
+            {
+                throw new ArgumentNullException(nameof(detailsProvider));
+            }
+
+            DetailsProvider = detailsProvider;
+            _cacheEntryFactory = CreateCacheEntry;
+            _messageProvider = messageProvider;
```

Comment:
https://github.com/aspnet/Mvc/pull/4928/files#diff-c8fa65ecc8077ca10cd5f923c0082de2R62 among others. There's lots of virtual stuff